DexPriceAggregatorUniswapV3/contracts/DexPriceAggregatorUniswapV3.sol

```diff
@@ -0,0 +1,409 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "./interfaces/IUniswapV3Pool.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/utils/math/SafeCast.sol";
+
+import "./libraries/PoolAddress.sol";
+import "./libraries/OracleLibrary.sol";
+import "./interfaces/IDexPriceAggregator.sol";
+
+/// @title DexPriceAggregatorUniswapV3
+/// @notice A DexPriceAggregator sourcing prices from Uniswap V3.
+///         Provides the minimum output between an asset's "spot" price and TWAP from the last n seconds.
+///         The "spot" price is always the last block's ending price.
+contract DexPriceAggregatorUniswapV3 is IDexPriceAggregator, Ownable {
+    address public immutable uniswapV3Factory;
+    address public immutable weth;
+    uint24 public immutable defaultPoolFee;
+
+    mapping(bytes32 => address) public overriddenPoolForRoute;
+
+    constructor(
+        address _uniswapV3Factory,
+        address _weth,
+        uint24 _defaultPoolFee
+    ) {
+        require(
+            _uniswapV3Factory != address(0),
+            "Can't set UniswapV3Factory address as 0"
+        );
+        require(_weth != address(0), "Can't set WETH address as 0");
+        uniswapV3Factory = _uniswapV3Factory;
+        weth = _weth;
+        defaultPoolFee = _defaultPoolFee;
+    }
+
+    /********************
+     * Oracle functions *
+     ********************/
+
+    /// @inheritdoc IDexPriceAggregator
+    function assetToAsset(
```

Comment:
-  _tokenIn and _tokenOut should not be the zero address
- _tokenIn should not be equal _tokenOut